False positive: `too-many-statements-in-try-clause` triggers on try-finally clause

### Summary

https://docs.astral.sh/ruff/rules/too-many-statements-in-try-clause/

Sometimes, it is easier to write `try-finally` clauses instead of context managers. Especially if the object is not a context manager, or we need to do other clean-up steps than `session.close()`. Anyway, this rule should not be triggered on a try-finally clause

```py
session = ...
try:
    print()
    print()
    print()
    print()
    print()
    print()
finally:
    session.close()
```

[Playground](https://play.ruff.rs/df009f98-b97c-4d8d-a74f-de06f20a42d8)


### Version

0.15.14

Thanks for the feedback on this preview rule! I'm not sure I follow why the rule shouldn't apply in this case. It sounds more like you may want to use a `noqa` comment in this instance (if you like the rule in general) or simply ignore the rule. Is this a common pattern in your code base?

This also seems consistent with pylint:

```console
$ uvx pylint try.py --load-plugins pylint.extensions.broad_try_clause --disable=all --enable=too-many-try-statements
************* Module try
try.py:2:0: W0717: try clause contains 6 statements, expected at most 1 (too-many-try-statements)
```

--- a/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
+++ b/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
@@ -1,0 +1,60 @@
+# `too-many-statements-in-try-clause` (`PLW0717`)
+
+```toml
+[lint]
+preview = true
+select = ["PLW0717"]
+```
+
+## Diagnostic range
+
+Avoid marking the entire `try` statement:
+
+```py
+# snapshot: too-many-statements-in-try-clause
+try:
+    call1()
+    call2()
+    call3()
+    call4()
+    call5()
+    call6()
+except:
+    ...
+```
+
+```snapshot
+error[PLW0717]: Try clause contains too many statements (6 > 5)
+  --> src/mdtest_snippet.py:2:1
+   |
+ 2 | / try:
+ 3 | |     call1()
+ 4 | |     call2()
+ 5 | |     call3()
+ 6 | |     call4()
+ 7 | |     call5()
+ 8 | |     call6()
+ 9 | | except:
+10 | |     ...
+   | |_______^
+   |
+```
+
+## Context managers
+
+Avoid emitting a diagnostic when the `try` statement is being used like a context manager, i.e. it
+catches no exceptions and just ensures that some kind of cleanup is run in a `finally` clause:
+
+```py
+# TODO false positive
+# error: [too-many-statements-in-try-clause]
+try:
+    call1()
+    call2()
+    call3()
+    call4()
+    call5()
+    call6()
+finally:
+    cleanup()
+```

--- a/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
+++ b/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
@@ -38,8 +38,6 @@ Avoid emitting a diagnostic when the `try` statement is being used like a contex
 catches no exceptions and just ensures that some kind of cleanup is run in a `finally` clause:
 
 ```py
-# TODO false positive
-# error: [too-many-statements-in-try-clause]
 try:
     call1()
     call2()

--- a/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
+++ b/crates/ruff_linter/resources/mdtest/pylint/too-many-statements-in-try-clause.md
@@ -25,19 +25,11 @@ except:
 
 ```snapshot
 error[PLW0717]: Try clause contains too many statements (6 > 5)
-  --> src/mdtest_snippet.py:2:1
-   |
- 2 | / try:
- 3 | |     call1()
- 4 | |     call2()
- 5 | |     call3()
- 6 | |     call4()
- 7 | |     call5()
- 8 | |     call6()
- 9 | | except:
-10 | |     ...
-   | |_______^
-   |
+ --> src/mdtest_snippet.py:2:1
+  |
+2 | try:
+  | ^^^
+  |
 ```
 
 ## Context managers

--- a/crates/ruff_linter/src/checkers/ast/analyze/statement.rs
+++ b/crates/ruff_linter/src/checkers/ast/analyze/statement.rs
@@ -1341,8 +1341,7 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
             if checker.is_rule_enabled(Rule::TooManyStatementsInTryClause) {
                 pylint::rules::too_many_try_statements(
                     checker,
-                    stmt,
-                    body,
+                    try_stmt,
                     checker.settings().pylint.max_statements_in_try,
                 );
             }

--- a/crates/ruff_linter/src/rules/pylint/rules/too_many_try_statements.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/too_many_try_statements.rs
@@ -1,5 +1,5 @@
 use ruff_macros::{ViolationMetadata, derive_message_formats};
-use ruff_python_ast::Stmt;
+use ruff_python_ast::StmtTry;
 use ruff_text_size::{Ranged, TextLen, TextRange};
 
 use crate::Violation;
@@ -89,18 +89,23 @@ impl Violation for TooManyStatementsInTryClause {
 /// W0717
 pub(crate) fn too_many_try_statements(
     checker: &Checker,
-    stmt: &Stmt,
-    body: &[Stmt],
+    try_stmt: &StmtTry,
     max_statements: usize,
 ) {
-    let statements = num_statements(body);
+    // Ignore cases with only `finally` clauses and no `except` handlers. These can't accidentally
+    // catch the wrong exception and are instead being used more like context managers.
+    if try_stmt.handlers.is_empty() {
+        return;
+    }
+
+    let statements = num_statements(&try_stmt.body);
     if statements > max_statements {
         checker.report_diagnostic(
             TooManyStatementsInTryClause {
                 statements,
                 max_statements,
             },
-            TextRange::at(stmt.start(), "try".text_len()),
+            TextRange::at(try_stmt.start(), "try".text_len()),
         );
     }
 }

--- a/crates/ruff_linter/src/rules/pylint/rules/too_many_try_statements.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/too_many_try_statements.rs
@@ -1,6 +1,6 @@
 use ruff_macros::{ViolationMetadata, derive_message_formats};
 use ruff_python_ast::Stmt;
-use ruff_python_ast::identifier::Identifier;
+use ruff_text_size::{Ranged, TextLen, TextRange};
 
 use crate::Violation;
 
@@ -100,7 +100,7 @@ pub(crate) fn too_many_try_statements(
                 statements,
                 max_statements,
             },
-            stmt.identifier(),
+            TextRange::at(stmt.start(), "try".text_len()),
         );
     }
 }

